include/aws/cal/symmetric_cipher.h

```diff
@@ -0,0 +1,263 @@
+ *
+ * respectively.
+ *
+ * If they are set, that key and iv will be copied internally and used by the cipher.
+ *
+ * Returns NULL on failure. You can check aws_last_error() to get the error code indicating the failure cause.
+ */
+AWS_CAL_API struct aws_symmetric_cipher *aws_aes_ctr_256_new(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *key,
+    const struct aws_byte_cursor *iv);
+
+/**
+ * Creates an instance of AES GCM with 256-bit key.
+ * If key, iv are NULL, they will be generated internally.
+ * You can get the generated key and iv back by calling:
+ *
+ * aws_symmetric_cipher_get_key() and
+ * aws_symmetric_cipher_get_initialization_vector()
+ *
+ * respectively.
+ *
+ * If they are set, that key and iv will be copied internally and used by the cipher.
+ *
+ * If tag and aad are set they will be copied internally and used by the cipher.
+ * decryption_tag would most likely be used for a decrypt operation to detect tampering or corruption.
+ * The Tag for the most recent encrypt operation will be available in:
+ *
+ * aws_symmetric_cipher_get_tag()
+ *
+ * If aad is set it will be copied and applied to the cipher.
+ *
+ * Returns NULL on failure. You can check aws_last_error() to get the error code indicating the failure cause.
+ */
+AWS_CAL_API struct aws_symmetric_cipher *aws_aes_gcm_256_new(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *key,
+    const struct aws_byte_cursor *iv,
+    const struct aws_byte_cursor *aad,
+    const struct aws_byte_cursor *decryption_tag);
+
+/**
+ * Creates an instance of AES Keywrap with 256-bit key.
+ * If key is NULL, it will be generated internally.
+ * You can get the generated key back by calling:
+ *
+ * aws_symmetric_cipher_get_key()
+ *
+ * If key is set, that key will be copied internally and used by the cipher.
+ *
+ * Returns NULL on failure. You can check aws_last_error() to get the error code indicating the failure cause.
+ */
+AWS_CAL_API struct aws_symmetric_cipher *aws_aes_keywrap_256_new(
+    struct aws_allocator *allocator,
+    const struct aws_byte_cursor *key);
+
+/**
+ * Cleans up internal resources and state for cipher and then deallocates it.
+ */
+AWS_CAL_API void aws_symmetric_cipher_destroy(struct aws_symmetric_cipher *cipher);
+
+/**
+ * Encrypts the value in to_encrypt and writes the encrypted data into out.
+ * If out is dynamic it will be expanded. If it is not, and out is not large enough to handle
+ * the encrypted output, the call will fail. If you're trying to optimize to use a stack based array
+ * or something, make sure it's at least as large as the size of to_encrypt + an extra BLOCK to account for
+ * padding etc...
+ *
+ * returns AWS_OP_SUCCESS on success. Call aws_last_error() to determine the failure cause if it returns
+ * AWS_OP_ERR;
+ */
+AWS_CAL_API int aws_symmetric_cipher_encrypt(
+    struct aws_symmetric_cipher *cipher,
+    const struct aws_byte_cursor *to_encrypt,
```

Comment:
If a byte-cursor is non-optional pass it by value

```suggestion
   struct aws_byte_cursor to_encrypt,
```

I know we haven't been 100% consistent about this in our code base, but it's better by value. It's obviously non-null, and you can "transform" it like aws_byte_cursor_from_buf() or aws_byte_cursor_from_c_str() without needing a tmp variable